packages/frontend/component/src/components/page-list/page-list.tsx

```diff
@@ -1,69 +1,120 @@
-import { Provider, useAtom, useAtomValue, useSetAtom } from 'jotai';
-import { useHydrateAtoms } from 'jotai/utils';
 import {
   type ForwardedRef,
   forwardRef,
-  type MouseEventHandler,
   type PropsWithChildren,
-  type ReactNode,
   useCallback,
   useEffect,
   useImperativeHandle,
-  useMemo,
   useRef,
 } from 'react';
 
-import { Checkbox, type CheckboxProps } from '../../ui/checkbox';
 import { useHasScrollTop } from '../app-sidebar/sidebar-containers/use-has-scroll-top';
 import { PageGroup } from './page-group';
+import { PageListTableHeader } from './page-header';
 import * as styles from './page-list.css';
 import {
   pageGroupsAtom,
-  pageListHandlersAtom,
   pageListPropsAtom,
-  pagesAtom,
+  PageListProvider,
   selectionStateAtom,
-  showOperationsAtom,
-  sorterAtom,
+  useAtom,
+  useAtomValue,
+  useSetAtom,
 } from './scoped-atoms';
 import type { PageListHandle, PageListProps } from './types';
-import { ColWrapper, type ColWrapperProps, stopPropagation } from './utils';
 
 /**
  * Given a list of pages, render a list of pages
  */
 export const PageList = forwardRef<PageListHandle, PageListProps>(
-  function PageListHandle(props, ref) {
+  function PageList(props, ref) {
     return (
-      <Provider>
-        <PageListInner {...props} handleRef={ref} />
-      </Provider>
+      // push pageListProps to the atom so that downstream components can consume it
+      // this makes sure pageListPropsAtom is always populated
+      // @ts-expect-error fix type issues later
+      <PageListProvider initialValues={[[pageListPropsAtom, props]]}>
+        <PageListInnerWrapper {...props} handleRef={ref}>
+          <PageListInner {...props} />
+        </PageListInnerWrapper>
+      </PageListProvider>
     );
   }
 );
 
-const PageListInner = ({
+// when pressing ESC or double clicking outside of the page list, close the selection mode
+// todo: use jotai-effect instead but it seems it does not work with jotai-scope?
+const usePageSelectionStateEffect = () => {
+  const [selectionState, setSelectionActive] = useAtom(selectionStateAtom);
+  useEffect(() => {
+    if (
+      selectionState.selectionActive &&
+      selectionState.selectable === 'toggle'
+    ) {
+      const startTime = Date.now();
+      const dblClickHandler = (e: MouseEvent) => {
+        if (Date.now() - startTime < 200) {
+          return;
+        }
+        const target = e.target as HTMLElement;
+        // skip if event target is inside of a button or input
+        // or within a toolbar (like page list floating toolbar)
+        if (
+          target.tagName === 'BUTTON' ||
+          target.tagName === 'INPUT' ||
+          (e.target as HTMLElement).closest('button, input, [role="toolbar"]')
+        ) {
+          return;
+        }
+        setSelectionActive(false);
+      };
+
+      const escHandler = (e: KeyboardEvent) => {
+        if (Date.now() - startTime < 200) {
+          return;
+        }
+        if (e.key === 'Escape') {
```

Comment:
Can we use enum or public lib to avoid typo maybe ?

Reply:
there is one ts lib that can have type checking on keyboardevent.key field.
i am not aware if there is a more broadly used library that have a enum mapping.

i think it should be fine without introducing an enum here since KeyboardEvent.key is a w3c standard and we shall refer to the mdn doc before using it https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values

Reply:
ok